# Exploring Ebay Car Sales Data

## Introduction

### Preliminary

This Notebook is the conclusion of the ***Numpys and Pandas Fundamental courses*** from [dataquest.io](dataquest.io). It is a guided project whose aim is to use all the techniques and skills learnt during the course. Nevertheless we are dealing with real-word data (in our case a [dataset](https://github.com/benny233-debug/Exploring-Ebay-Car-Sales-Data/blob/master/autos.csv) scrapped from eBay Kleinanzeigen)
> I am sometimes using blockquotes as this one, meanign that for the rest of the project I am quoting some elements given by dataquest. For the sake of simplicity and clarity, I estimated that they did not need any reformulation and were immediately usable and convenient for me and the reader.

### Context

>In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The dataset was originally scraped and uploaded to Kaggle by user orgesleka.
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://github.com/benny233-debug/Exploring-Ebay-Car-Sales-Data/blob/master/autos.csv).

>We've made a few modifications from the original dataset:
- We sampled 50,000 data points from the full dataset, to ensure your code runs quickly in our hosted environment
- We dirtied the dataset a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

> The data dictionary provided with data is as follows:
- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online


The aim of this project is to **clean** the data and **analyze** the included used car listings

## Cleaning the Data

### reading the data

In [1]:
#importing the librairies
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv',encoding='Latin-1')

Now let's print some information about our data set as well as the first rows of our Data set.

In [2]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

As showed above, our data set contains 50000 rows and 20 columns. 15 of the columns contain *object* meaning essentially string types and the rest are integer. Among the 15 columns containing objects, we can see that some of them might need to be converted to integer like for instance the price and the odometer colums.  

Another aspect that we will want to deal with is that there are five columns that contain null values :
- vehicleType
- gearbox
- model 
- fuelType
- notRepairedDamage  

Notice that for none of the previous columns more than 20% of the rows is missing.

Eventually, one can notice that the column names use *camelCase* and not the *snakecase* convention which is normally preferred in Python. So we will have to rename the columns following the latter. We will start first we this step.


### Renaming the columns

In [4]:
column_names=autos.columns
print(column_names)
autos=autos.rename(columns={'dateCrawled':'date_crawled','offerType':'offer_type','vehicleType':'vehicle_type',
              'yearOfRegistration':'registration_year','powerPS':'power_ps','monthOfRegistration':'registration_month',
             'fuelType':'fuel_type','notRepairedDamage':'unrepaired_damage','dateCreated':'ad_created',
              'nrOfPictures':'nr_of_pictures','postalCode':'postal_code','lastSeen':'last_seen','odometer':'odometer_km'})

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model odometer_km  registration_month fuel_type  \
0    manuell       158  andere   150,000km                   3       lpg   
1  automatik       286     7er   150,000km                   6    benzin   
2    manuell       102    golf    70,000km                   7    benzin   
3  automatik        71  fortwo    70,000km                   6    benzin   
4    manuell         0   focus   150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

So now we have renammed all the names of our colums so they are following the *snakecase* convention

### inital exploration and cleaning

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

       odometer_km  registration_month fuel_type       brand  \
count        50000        50000.000000     45518       50000   
unique          13                 NaN         7          40   
top      150,000km                 NaN    benzin  volkswagen   
freq         32424                 NaN     30107       10687   
mean           NaN            5.723360       NaN         NaN   
std            NaN            3.711984       NaN         NaN   
min            NaN            0.000000       NaN         NaN   
25%            NaN            3.000000       NaN         NaN   
50%            NaN            6.000000       NaN         NaN   
75%            NaN            9.000000       NaN         NaN   
max            NaN           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count              40171                50000         50000.0  50000.000000   
unique                 2                   76             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq               35232                 1946             NaN           NaN   
mean                 NaN                  NaN             0.0  50813.627300   
std                  NaN                  NaN             0.0  25779.747957   
min                  NaN                  NaN             0.0   1067.000000   
25%                  NaN                  NaN             0.0  30451.000000   
50%                  NaN                  NaN             0.0  49577.000000   
75%                  NaN                  NaN             0.0  71540.000000   
max                  NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
abtest                50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gearbox               47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer_km           50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
nr_of_pictures        50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

With the Dataframe.describe() method that we have used above we can see that : 
- the 'nr_of_pictures' column only contains 0 so we will drop it.
- several columns only have two values : offer_type, seller, abtest, gearbox, unrepaired_damage. By using the *Series.value_counts() method we realize that both seller and offer_type have 49999 over 50000 the same value : respectively "privat" and "Angebot" so we might delete these columns and they don't add valuable information.
- The gearbox and unrepaired_damage column must be kept as they bring useful information for a second-hand car sell.
- The abtest should be investigated.It has two values (test or control) that are almost evenly distributed. The AB tests are made normally be the companies when they want to try for instance two marketing campaings or two variations of a feature on two different targets and see if one or the other is more succesful than the other and then draw conclusions. So my guess is that eBay made a test to implement a new feature for instance.
- The price and odometer columns are stored as string with a special format for both of them. On the first one the format is stored with '$' before the number so we should clean this and store it as an integer and same for odometer where the 'km' should be remove from the string and then convert the latter as an integer

Also as we are dealing with Sell-ads, we can infer that all the missing values for the columns given above are due to the fact that the sellers did not enter all the details concerning their car as they are not mandatory to publish an add.



In [8]:
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis=1)

In [9]:
autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(float)


In [10]:
autos['price'].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [11]:
autos['odometer_km']=autos['odometer_km'].str.replace('km','').str.replace(',','').astype(float)


### exploring the odometer_km and price colums
#### odometer

In [12]:
autos['odometer_km'].unique()


array([ 150000.,   70000.,   50000.,   80000.,   10000.,   30000.,
        125000.,   90000.,   20000.,   60000.,    5000.,  100000.,   40000.])

In [13]:
autos['odometer_km'].unique().shape


(13,)

In [14]:
autos['odometer_km'].describe()


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

The odometer_km colomun has 13 diferent values ranging from 5000 to 150000.
The values in the columns all seem reasonable as we are dealing with kilometers for second-hand cars.  
Using the Series.describe() and Series.value_counts() methods we are able to see the distribution of our data. It turns out that more than 50% of the Data has the **150000** value and than more than 75% of the data are distributed between **125000** and **150000**.
We might supose these values mean for instance more than X km or less than Y km. For instance the *5000* value could be a range from 0 to 5000km, the 10000 a range from 5000 to 10000 and eventually the 150000 meaning (between 125k and 150k).
There no outliers in this series

#### Price

In [16]:
autos['price'].unique().shape


(2357,)

In [17]:
autos['price'].describe()


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

The price column seems to have uncorrect data or outliers.
First we can notice that they are 2357 different values in this column.  
The min value seems to be wrong as the price of a second-hand car should be reasonably greater than 0. ***Unless that that the site allows the user to set a price = 0, if the latter wants to give away their car ***.
The max value also seems incorrect as a 100000000-euro price is non-sense.

In [18]:
autos['price'].value_counts().sort_index(ascending=True).head(10)


0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64

In [19]:
autos['price'].value_counts().sort_index(ascending=False).head(10)


99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64

Thanks to the operations made above, it is said that **1421** persons fixed the price of their car to **0**. So our guess that they are people that want to get ride of their car might hold. Besides 156 persons fixed their price to 1€ and a few others to very symobolical values for a cars (between 1 and 100€).
On the contrary for the "top" outliers, they are very less frequent, none of the senseless values (above 1.3 Million € given that the most expensive cars seldomly cost over 1 or 2 million euros. The most expensive car in the world , the ***Bugatti Noire*** was sold around 11M€ )  
Besides, it is very unlikely that the owner of "Super Car" would sell theirs on Ebay KleineAnzeigen so will remove all the rows who have a price over 35000€. But we will keep for now the rows with a low or null price.

In [20]:
autos_0 = autos[autos["price"]==0]
autos = autos[autos["price"]<350000]

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

### exploring the dates colums 
#### date_crawled

In [22]:
print(autos['date_crawled'].str[:10].min())
print(autos['date_crawled'].str[:10].max())

autos['date_crawled'].str[:10].describe()

2016-03-05
2016-04-07


count          49985
unique            34
top       2016-04-03
freq            1934
Name: date_crawled, dtype: object

In [23]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025388
2016-03-06    0.013944
2016-03-07    0.035971
2016-03-08    0.033270
2016-03-09    0.033210
2016-03-10    0.032130
2016-03-11    0.032490
2016-03-12    0.036771
2016-03-13    0.015565
2016-03-14    0.036631
2016-03-15    0.033990
2016-03-16    0.029509
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037491
2016-03-22    0.032910
2016-03-23    0.032390
2016-03-24    0.029109
2016-03-25    0.031750
2016-03-26    0.032490
2016-03-27    0.031029
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033630
2016-03-31    0.031910
2016-04-01    0.033810
2016-04-02    0.035411
2016-04-03    0.038692
2016-04-04    0.036491
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

There are 34 unique values in the *date_crawled* column corresponding to the dates between the 5th of March and the 7th of April 2016. We can see the values are mostly evenly distributed : ranging from 1.3% to 3.7% for each day except the last two days (April the 6th and 7th which account both for less than 1%.) 


#### ad_created

In [24]:
print(autos['ad_created'].str[:10].min())
print(autos['ad_created'].str[:10].max())

autos['ad_created'].str[:10].describe()

2015-06-11
2016-04-07


count          49985
unique            76
top       2016-04-03
freq            1946
Name: ad_created, dtype: object

In [25]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index().head(50)

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
2015-12-30    0.000020
2016-01-03    0.000020
2016-01-07    0.000020
2016-01-10    0.000040
2016-01-13    0.000020
2016-01-14    0.000020
2016-01-16    0.000020
2016-01-22    0.000020
2016-01-27    0.000060
2016-01-29    0.000020
2016-02-01    0.000020
2016-02-02    0.000040
2016-02-05    0.000040
2016-02-07    0.000020
2016-02-08    0.000020
2016-02-09    0.000040
2016-02-11    0.000020
2016-02-12    0.000060
2016-02-14    0.000040
2016-02-16    0.000020
2016-02-17    0.000020
2016-02-18    0.000040
2016-02-19    0.000060
2016-02-20    0.000040
2016-02-21    0.000060
2016-02-22    0.000020
2016-02-23    0.000080
2016-02-24    0.000040
2016-02-25    0.000060
2016-02-26    0.000040
2016-02-27    0.000120
2016-02-28    0.000200
2016-02-29    0.000160
2016-03-01    0.000100
2016-03-02    0.000100
2016-03-03    0.000860
2016-03-04    0.001440
2016-03-05    0.023047
2016-03-06 

There are 76 unique values in the *ad_created* column corresponding to dates from the 11th of June 2015 to the the 7th of April 2016. In facts all the values before March the 5th (i-e the earliest date in the *date_crawled* column) are negligeable as they barely account for 0.01 %. For the rest, the values seem to be evenly distributed.

#### last_seen

In [26]:
print(autos['last_seen'].str[:10].min())
print(autos['last_seen'].str[:10].max())

autos['last_seen'].str[:10].describe()

2016-03-05
2016-04-07


count          49985
unique            34
top       2016-04-06
freq           11046
Name: last_seen, dtype: object

In [27]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015885
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015745
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018586
2016-03-24    0.019566
2016-03-25    0.019206
2016-03-26    0.016965
2016-03-27    0.016005
2016-03-28    0.020846
2016-03-29    0.022327
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023107
2016-04-02    0.024887
2016-04-03    0.025368
2016-04-04    0.024627
2016-04-05    0.124277
2016-04-06    0.220986
2016-04-07    0.130959
Name: last_seen, dtype: float64

There are 34 unique values in the *date_crawled* column corresponding to the dates between the 5th of March and the 7th of April 2016. The values are not evenly distributed. Indeed, the last three dates (5th,6th and 7th of April) account for nearly half of the values. It seems logical since people are more likely to click on recent ads (relatively to the scrapping date 7th April) than old ones. In the wake of this remark, we can notice that the oldest dates account for less than 1% each one.


#### registration_year

In [28]:
autos['registration_year'].describe()

count    49985.000000
mean      2005.075503
std        105.728207
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [29]:
autos['registration_year'].value_counts(normalize=True,dropna=False).sort_index()

1000    0.000020
1001    0.000020
1111    0.000020
1500    0.000020
1800    0.000040
1910    0.000180
1927    0.000020
1929    0.000020
1931    0.000020
1934    0.000040
1937    0.000080
1938    0.000020
1939    0.000020
1941    0.000040
1943    0.000020
1948    0.000020
1950    0.000060
1951    0.000040
1952    0.000020
1953    0.000020
1954    0.000040
1955    0.000040
1956    0.000100
1957    0.000040
1958    0.000080
1959    0.000140
1960    0.000660
1961    0.000120
1962    0.000080
1963    0.000180
          ...   
2001    0.054056
2002    0.050675
2003    0.054556
2004    0.054756
2005    0.060318
2006    0.054156
2007    0.046094
2008    0.044633
2009    0.041953
2010    0.031950
2011    0.032690
2012    0.026468
2013    0.016125
2014    0.013304
2015    0.007982
2016    0.026308
2017    0.029049
2018    0.009823
2019    0.000060
2800    0.000020
4100    0.000020
4500    0.000020
4800    0.000020
5000    0.000080
5911    0.000020
6200    0.000020
8888    0.000020
9000    0.0000

The two previous operations made us realize that there are *outliers* in this column for instance : 1000 or 9999. (see above)
As the first Ford T was commercialized in 1908 we will estimate that the values before 1927 are not reasonable and neither are the values after 2016 (scrapping year)

In [30]:
autos['registration_year'][autos['registration_year'].between(1927,2016)].describe()

count    48006.000000
mean      2002.823126
std          7.195358
min       1927.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [31]:
autos['registration_year'][autos['registration_year'].between(1927,2016)].value_counts(normalize=True).head(15).sum()

0.7731741865600131

In [32]:
autos['registration_year'][autos['registration_year'].between(1927,2016)].value_counts(normalize=True).sort_index().tail(5).sum()

0.09390492855059784

From the previous operations, we can see that more than 75% of our vehicles are from 1999 and over, meaning that the majority of the ads concern relatively "recent" cars.  
We can be more precise using the *.series.head(15).sum() method.* : vehicles between 1997 and 2011 (included) account for **77%** of our data and those between 2012 and 2016 : **9.4%**

### Exploring Price by Brand

In [33]:
autos['brand'].describe()

count          49985
unique            40
top       volkswagen
freq           10684
Name: brand, dtype: object

In [34]:
autos['brand'].unique().shape

(40,)

In [35]:
autos['brand'].value_counts(normalize=True).head(10)

volkswagen       0.213744
opel             0.109233
bmw              0.108593
mercedes_benz    0.094688
audi             0.085686
ford             0.069561
renault          0.048094
peugeot          0.029129
fiat             0.026148
seat             0.018826
Name: brand, dtype: float64

Over the 40 uniques brand that are in our column, we will choose to perform the aggregation only on 10 brands (see above). To make this choice, we used the used the *.value_counts()* method and summed up the 10 most frequent brands. It turns out that they account for 80% of the values.  

We will try now to get the average price for each of this brand.

In [36]:
ten_brands=autos['brand'].value_counts(normalize=True).head(10).index

In [37]:
avg_price_ten_brands = {}
for b in ten_brands:
    avg = autos['price'][autos['brand']==b].mean()
    avg_price_ten_brands[b]= round(avg)
    

In [38]:
print(avg_price_ten_brands)

{'volkswagen': 5159, 'opel': 2846, 'bmw': 8027, 'mercedes_benz': 8390, 'audi': 8966, 'ford': 3627, 'renault': 2351, 'peugeot': 3011, 'fiat': 2698, 'seat': 4219}


Looking at our previously created dictionnary we can conclude that :
- The average price of a  volkswagen (which is the most frequent brand to be sold) is 5159€
- The lowest average prices are :  2351€ for a renault, 2698€ for a fiat, and 2848€ for an opel. 
- The highest average prices are : 8966€ for an audi, 8390€ for a Mercedes and 8027€ for a BMW.

Now let's have a look at the average mileage for theses 10 brands and to eventually agregate both average prices and average mileages by brand

In [39]:
avg_mileage_ten_brands = {}
for b in ten_brands:
    mil = autos['odometer_km'][autos['brand']==b].mean()
    avg_mileage_ten_brands[b]= round(mil)

In [40]:

avg_price_series = pd.Series(avg_price_ten_brands)
avg_mileage_series = pd.Series(avg_mileage_ten_brands)

df = pd.DataFrame(avg_price_series,columns=['mean_price'])
df['mean_mileage']= avg_mileage_series

In [41]:
df

mean_price  mean_mileage
audi                 8966        129644
bmw                  8027        132518
fiat                 2698        117012
ford                 3627        124153
mercedes_benz        8390        130882
opel                 2846        129295
peugeot              3011        127352
renault              2351        128224
seat                 4219        122062
volkswagen           5159        128949

In [42]:
df.describe()

mean_price   mean_mileage
count    10.000000      10.000000
mean   4929.400000  127009.100000
std    2576.965364    4660.057926
min    2351.000000  117012.000000
25%    2887.250000  124952.750000
50%    3923.000000  128586.500000
75%    7310.000000  129556.750000
max    8966.000000  132518.000000

With the previous dataframe, we can conclude that  there is no visible link between the mean price and the mean mileage as for the top 10 brands, they all have mean mileages relatively close to the average mean mileage (127 000km) and the standard deviation is equal to 4660 so the ratio std/mean = 3,7%

In [43]:
4660/127000

0.036692913385826774